### PR TITLE
Add information regarding the drop option in the linear model exercise

### LINKED ISSUE #276: Feedback on Linear Model Chapter
This MOOC is really great. Below are a few suggestions after a first partial read.

Chapter [Linear regression using scikit-learn](https://inria.github.io/scikit-learn-mooc/python_scripts/linear_regression_in_sklearn.html):

- Mention that the strategy for finding the best parameters is called (ordinary) **least squares**.

[Linear regression with non-linear link between data and target](https://inria.github.io/scikit-learn-mooc/python_scripts/linear_regression_non_linear_link.html)

- The word *link* is unfortunate as it might get confused with the *link* function of GLMs.
- As soon as v1.0 of scikit-learn arrives, mention splines with `SplineTransformer`.
- Speeking about *interactions* is a bit misleading in a setting with only one single feature. But I would really love to see an example that specifies 2 interaction terms: Given x1 to x3, create x1, x2, x3, x1*x2, x1*x3 (not x2*x3 and not x**2 terms either).

[Regularization of linear regression model](https://inria.github.io/scikit-learn-mooc/python_scripts/linear_models_regularization.html)

- Remove convergence warnings of Ridge (other solver?)
- Maybe a word about categorical variables and the `drop` parameter in OHE?
- "Let’s consider the case where features have an identical data dispersion". The term "dispersion" seems unusual to me. How about "features live on the same scale/dimension like both are times measured in microseconds."
   Later on: "our model will boost the weights of features with small dispersion" => It is about larger/smaller (floating point/machine) numbers, not about dispersion.
- What about scaling of OHE categorical features?

[Main take-away](https://inria.github.io/scikit-learn-mooc/linear_models/linear_models_module_take_away.html)

- Again the word "link".
- Mention that (manual) feature engineering and feature selection is (by far) the greatest impact/tuning factor for linear models (and therefore against over- and underfitting).

### PR DESCRIPTION
**Pull request by glemaitre**
closes #276 

Introduce the parameter `drop` from the `OneHotEncoder` through the exercise presenting the problem of collinear features.

### CODE REVIEW

File: python_scripts/linear_models_sol_04.py [hunk 1]
```diff
@@ -136,3 +136,101 @@
...
+    X_trans,
+    columns=encoder.get_feature_names_out(input_features=single_feature),
+)
+X_trans
+
+# %% [markdown] tags=["solution"]
+#
+# Here, we see that the encoded category "CentralAir_N" is the opposite of the
+# encoded category "CentralAir_Y". Therefore, we observe that using a
+# `OneHotEncder` creates features having the problematic pattern observe
```
Comment (ogrisel) [inline python_scripts/linear_models_sol_04.py:191]:
```suggestion
# `OneHotEncoder` creates two features having the problematic pattern observed
```

File: python_scripts/linear_models_sol_04.py [hunk 2]
```diff
@@ -136,3 +136,101 @@
...
+# earlier in this exercise. Training a linear model, only one of the feature
+# is therefore informative because they carry the same information.
```
Comment (ogrisel) [inline python_scripts/linear_models_sol_04.py:193]:
```suggestion
# earlier in this exercise. Training a linear regression model on such a
# of one-hot encoded binary feature can therefore lead to numerical
# problems, especially without regularization. Furthermore, the two one-hot
# feature are redundant as they encode exactly the same information in
# opposite ways.
```

File: python_scripts/linear_models_sol_04.py [hunk 3]
```diff
@@ -136,3 +136,101 @@
...
+#
+# Using regularization will thus help to overcome the issues that we spotted
```
Comment (ogrisel) [inline python_scripts/linear_models_sol_04.py:195]:
```suggestion
# Using regularization will helps to overcome the numerical issues that we highlighted
```

File: python_scripts/linear_models_sol_04.py [hunk 4]
```diff
@@ -136,3 +136,101 @@
...
+# earlier in this exercise.
+#
+# Another strategy is to arbitrary drop one of the encoded categories.
```
Comment (ogrisel) [inline python_scripts/linear_models_sol_04.py:198]:
```suggestion
# Another strategy is to arbitrarily drop one of the encoded categories.
```

File: python_scripts/linear_models_sol_04.py [hunk 5]
```diff
@@ -136,3 +136,101 @@
...
+    columns=encoder.get_feature_names_out(input_features=single_feature),
+)
+X_trans
+
+# %% [markdown] tags=["solution"]
+#
+# We see that only the second column of the previous encoded data is kept.
+# This preprocessing is commonly used. However, you should notice that it
+# break feature symmetry and will impact the number of coefficients of the
+# model, their values, and thus their meaning.
```
Comment (ogrisel) [inline python_scripts/linear_models_sol_04.py:219]:
```suggestion
# Dropping one of the one-hot encoded column is a common practice,
# especially for binary categorical features. Note however that this breaks
# symmetry between categories and impacts the number of coefficients of the
# model, their values, and thus their meaning, especially when applying
# strong regularization.
```

File: python_scripts/linear_models_sol_04.py [hunk 6]
```diff
@@ -136,3 +136,101 @@
...
+#
+# Note that we can use this option is a machine-learning pipeline.
```
Comment (ogrisel) [inline python_scripts/linear_models_sol_04.py:221]:
```suggestion
# Let's finally illustrate how to use this option is a machine-learning pipeline:
```

**Comment by glemaitre** [comment]
Let me also rename the feature names as we previously discussed during a call to make it easier to read the coefficients.

File: python_scripts/linear_models_sol_04.py [hunk 7]
```diff
@@ -80,7 +92,17 @@
...
+feature_names = [
+    "Relevant feature #0",
+    "Relevant feature #1",
+    "Noisy feature #0",
+    "Noisy feature #1",
+    "Noisy feature #2",
+    "Repeated feature #0",
+    "Repeated feature #1",
+    "Repeated feature #0",
+    "Repeated feature #1",
```
Comment (ogrisel) [inline python_scripts/linear_models_sol_04.py:104]:
```suggestion
    "First repetition of feature #0",
    "First repetition of  feature #1",
    "Second repetition of  feature #0",
    "Second repetition of  feature #1",
```

File: python_scripts/linear_models_sol_04.py [hunk 8]
```diff
@@ -136,3 +158,110 @@
...
+
+# %% [markdown] tags=["solution"]
+#
+# Here, we see that the encoded category "CentralAir_N" is the opposite of the
+# encoded category "CentralAir_Y". Therefore, we observe that using a
+# `OneHotEncoder` creates two features having the problematic pattern observed
+# earlier in this exercise. Training a linear regression model on such a
+# of one-hot encoded binary feature can therefore lead to numerical
+# problems, especially without regularization. Furthermore, the two one-hot
+# feature are redundant as they encode exactly the same information in
```
Comment (ogrisel) [inline python_scripts/linear_models_sol_04.py:221]:
```suggestion
# features are redundant as they encode exactly the same information in
```

File: python_scripts/linear_models_sol_04.py [hunk 9]
```diff
@@ -136,3 +158,110 @@
...
+# opposite ways.
+#
+# Using regularization will helps to overcome the numerical issues that we highlighted
```
Comment (ogrisel) [inline python_scripts/linear_models_sol_04.py:224]:
```suggestion
# Using regularization helps to overcome the numerical issues that we highlighted
```

**Review by ogrisel — APPROVED** [review]
LGTM with the following remaining details. I will commit an merge.

**Pull request merged**